python/ray/data/dataset.py

```diff
@@ -2933,6 +2939,13 @@ def iter_torch_batches(
                 will be inferred from the tensor data.
             device: The device on which the tensor should be placed; if None, the Torch
                 tensor will be constructed on the CPU.
+            collate_fn: A function to convert a Numpy batch to a PyTorch tensor batch.
```

Comment:
Some of this could be explored as a follow-up, but I'm wondering if we should more closely align our `collate_fn` semantics to the semantics for the Torch `DataLoader`'s semantics, to make porting more straightforward? https://pytorch.org/docs/stable/data.html#working-with-collate-fn

A few questions along these lines:
1. When automatic batching is disabled (in our case, `batch_size=None`), should `collate_fn` get the entire block, or one row at a time, like the Torch `DataLoader`'s `collate_fn`? I'd vote for the former since our API is more batch-centric (which I think is a good thing), but want to know if there are any other thoughts.
2. When automatic batching is enabled (in our case, `batch_size is not None`), should the default `collate_fn` follow the same pattern as `DataLoader`'s default `collate_fn`, i.e. preserve the input data's structure more faithfully rather than always converting to dict-of-ndarray batches (primarily for the simple block case). I'm again thinking that the dict-of-ndarray narrow waist is a good one that we should keep.

Both (1) and (2) are discrepancies that users might run into when porting from `DataLoader`, so I thought that they should be called out, especially since we're adding `collate_fn` explicitly to make that porting process easier.

Reply:
Correct, the current implementation aligns with what I think we should do, this comment was a call-out that there will be a slight mismatch between the Torch Data `collate_fn` and our `collate_fn`, which might make porting a bit more difficult.

One action item would be calling out this discrepancy in our porting guide, or making those semantics extra clear in the `iter_torch_batches` docstring. I think that the type hint for `collate_fn` + the existing docstring is good enough, so maybe just an extra note in the porting guide/example (we have one of those, right?)

Reply:
We don't have one 🙃

Reply:
Ah shucks, I thought that we had a "port from `DataLoader` to Ray Datasets" in this example, but it looks like we don't: https://docs.ray.io/en/master/ray-air/examples/convert_existing_pytorch_code_to_ray_air.html#starting-with-ray-air-distribute-the-training

Ok, no action item here then! We'll just have to keep this in mind if/when we create that guide.